In [1]:
import gpxpy
import pandas as pd
import numpy as np

#### Read the GPS data along the route from Vijaya Nagar to Whitefield

In [20]:
gpx_file = open('vijayanagar_whitefield.gpx', 'r')

In [21]:
gpx = gpxpy.parse(gpx_file)

In [22]:
len(gpx.tracks)

1

In [23]:
len(gpx.tracks[0].segments)

1

In [24]:
len(gpx.tracks[0].segments[0].points)

488

##### There are 488 data points along the route.

In [25]:
data = gpx.tracks[0].segments[0].points

In [26]:
## Start Position
start = data[0]
## End Position
finish = data[-1]

In [27]:
print(start)

[trkpt:12.97177,77.54575@None@None]


In [28]:
print(finish)

[trkpt:12.96961,77.74988@None@None]


In [29]:
df = pd.DataFrame(columns=['lon', 'lat', 'alt', 'time'])

In [30]:
for point in data:
    df = df.append({'lon': point.longitude, 'lat' : point.latitude, 'alt' : point.elevation, 'time' : point.time}, ignore_index=True)

In [31]:
df.head()

lon       lat  alt  time
0  77.54575  12.97177  NaN   NaN
1  77.54669  12.97144  NaN   NaN
2  77.54669  12.97144  NaN   NaN
3  77.54656  12.97112  NaN   NaN
4  77.54641  12.97078  NaN   NaN

In [33]:
df=df.drop(['alt','time'],axis=1)

In [34]:
df.head()

lon       lat
0  77.54575  12.97177
1  77.54669  12.97144
2  77.54669  12.97144
3  77.54656  12.97112
4  77.54641  12.97078

In [35]:
df.tail()

lon       lat
483  77.74973  12.96877
484  77.74977  12.96899
485  77.74986  12.96951
486  77.74988  12.96961
487  77.74988  12.96961

#### Dummy Data is created for the location with latitude 77.54641 and longitude 12.97078

In [39]:
route_data = pd.read_csv('Public_Places_data.csv')

In [40]:
route_data

Public Place  Distance(km)  Duration(mins)  \
0   Women Police Station           0.8               5   
1         Police Station           1.0               7   
2                   Mall           1.1              10   
3                    NGO           1.1               8   
4   Traffic Control Room           1.3              14   
5         Busy Junctions           1.2              14   
6          Moving Patrol           0.9               9   
7          24/7 Hospital           1.4               8   
8                 Clinic           1.3               6   
9       24/7 Petrol Bunk           1.4              10   
10   Regular Petrol Bunk           1.2              12   

    User Rating(1-Poor,5-Excellent) open_at_night  Weightage  
0                                 5           Yes        1.0  
1                                 5           Yes        0.9  
2                                 4            No        0.7  
3                                 4            No        0.6  
4                                 3            No        0.8  
5                                 3            No        0.5  
6                                 5           Yes        0.9  
7                                 5           Yes        1.0  
8                                 4            No        0.7  
9                                 5           Yes        1.0  
10                                3            No        0.7

In [41]:
route_data["Distance_Rank"] = route_data["Distance(km)"].rank() 


In [44]:
route_data

Public Place  Distance(km)  Duration(mins)  \
0   Women Police Station           0.8               5   
1         Police Station           1.0               7   
2                   Mall           1.1              10   
3                    NGO           1.1               8   
4   Traffic Control Room           1.3              14   
5         Busy Junctions           1.2              14   
6          Moving Patrol           0.9               9   
7          24/7 Hospital           1.4               8   
8                 Clinic           1.3               6   
9       24/7 Petrol Bunk           1.4              10   
10   Regular Petrol Bunk           1.2              12   

    User Rating(1-Poor,5-Excellent) open_at_night  Weightage  Distance_Rank  \
0                                 5           Yes        1.0            1.0   
1                                 5           Yes        0.9            3.0   
2                                 4            No        0.7            4.5   
3                                 4            No        0.6            4.5   
4                                 3            No        0.8            8.5   
5                                 3            No        0.5            6.5   
6                                 5           Yes        0.9            2.0   
7                                 5           Yes        1.0           10.5   
8                                 4            No        0.7            8.5   
9                                 5           Yes        1.0           10.5   
10                                3            No        0.7            6.5   

    Duration_Rank  
0             1.0  
1             3.0  
2             7.5  
3             4.5  
4            10.5  
5            10.5  
6             6.0  
7             4.5  
8             2.0  
9             7.5  
10            9.0

In [43]:
route_data["Duration_Rank"] = route_data["Duration(mins)"].rank() 

In [48]:
route_data["Rating_Rank"] = route_data["User Rating(1-Poor,5-Excellent)"].rank(ascending=False) 

In [49]:
route_data

Public Place  Distance(km)  Duration(mins)  \
0   Women Police Station           0.8               5   
1         Police Station           1.0               7   
2                   Mall           1.1              10   
3                    NGO           1.1               8   
4   Traffic Control Room           1.3              14   
5         Busy Junctions           1.2              14   
6          Moving Patrol           0.9               9   
7          24/7 Hospital           1.4               8   
8                 Clinic           1.3               6   
9       24/7 Petrol Bunk           1.4              10   
10   Regular Petrol Bunk           1.2              12   

    User Rating(1-Poor,5-Excellent) open_at_night  Weightage  Distance_Rank  \
0                                 5           Yes        1.0            1.0   
1                                 5           Yes        0.9            3.0   
2                                 4            No        0.7            4.5   
3                                 4            No        0.6            4.5   
4                                 3            No        0.8            8.5   
5                                 3            No        0.5            6.5   
6                                 5           Yes        0.9            2.0   
7                                 5           Yes        1.0           10.5   
8                                 4            No        0.7            8.5   
9                                 5           Yes        1.0           10.5   
10                                3            No        0.7            6.5   

    Duration_Rank  Rating_Rank  
0             1.0          3.0  
1             3.0          3.0  
2             7.5          7.0  
3             4.5          7.0  
4            10.5         10.0  
5            10.5         10.0  
6             6.0          3.0  
7             4.5          3.0  
8             2.0          7.0  
9             7.5          3.0  
10            9.0         10.0

#### Calculate the score based on weighted average of Distance, Duration and Rating Rank

In [50]:
route_data['Score']=route_data[['Distance_Rank', 'Duration_Rank', 'Rating_Rank']].mean(axis=1)

In [51]:
route_data

Public Place  Distance(km)  Duration(mins)  \
0   Women Police Station           0.8               5   
1         Police Station           1.0               7   
2                   Mall           1.1              10   
3                    NGO           1.1               8   
4   Traffic Control Room           1.3              14   
5         Busy Junctions           1.2              14   
6          Moving Patrol           0.9               9   
7          24/7 Hospital           1.4               8   
8                 Clinic           1.3               6   
9       24/7 Petrol Bunk           1.4              10   
10   Regular Petrol Bunk           1.2              12   

    User Rating(1-Poor,5-Excellent) open_at_night  Weightage  Distance_Rank  \
0                                 5           Yes        1.0            1.0   
1                                 5           Yes        0.9            3.0   
2                                 4            No        0.7            4.5   
3                                 4            No        0.6            4.5   
4                                 3            No        0.8            8.5   
5                                 3            No        0.5            6.5   
6                                 5           Yes        0.9            2.0   
7                                 5           Yes        1.0           10.5   
8                                 4            No        0.7            8.5   
9                                 5           Yes        1.0           10.5   
10                                3            No        0.7            6.5   

    Duration_Rank  Rating_Rank     Score  
0             1.0          3.0  1.666667  
1             3.0          3.0  3.000000  
2             7.5          7.0  6.333333  
3             4.5          7.0  5.333333  
4            10.5         10.0  9.666667  
5            10.5         10.0  9.000000  
6             6.0          3.0  3.666667  
7             4.5          3.0  6.000000  
8             2.0          7.0  5.833333  
9             7.5          3.0  7.000000  
10            9.0         10.0  8.500000

#### Safe Zone Places ranking during Day 

In [52]:
route_data.sort_values('Score')

Public Place  Distance(km)  Duration(mins)  \
0   Women Police Station           0.8               5   
1         Police Station           1.0               7   
6          Moving Patrol           0.9               9   
3                    NGO           1.1               8   
8                 Clinic           1.3               6   
7          24/7 Hospital           1.4               8   
2                   Mall           1.1              10   
9       24/7 Petrol Bunk           1.4              10   
10   Regular Petrol Bunk           1.2              12   
5         Busy Junctions           1.2              14   
4   Traffic Control Room           1.3              14   

    User Rating(1-Poor,5-Excellent) open_at_night  Weightage  Distance_Rank  \
0                                 5           Yes        1.0            1.0   
1                                 5           Yes        0.9            3.0   
6                                 5           Yes        0.9            2.0   
3                                 4            No        0.6            4.5   
8                                 4            No        0.7            8.5   
7                                 5           Yes        1.0           10.5   
2                                 4            No        0.7            4.5   
9                                 5           Yes        1.0           10.5   
10                                3            No        0.7            6.5   
5                                 3            No        0.5            6.5   
4                                 3            No        0.8            8.5   

    Duration_Rank  Rating_Rank     Score  
0             1.0          3.0  1.666667  
1             3.0          3.0  3.000000  
6             6.0          3.0  3.666667  
3             4.5          7.0  5.333333  
8             2.0          7.0  5.833333  
7             4.5          3.0  6.000000  
2             7.5          7.0  6.333333  
9             7.5          3.0  7.000000  
10            9.0         10.0  8.500000  
5            10.5         10.0  9.000000  
4            10.5         10.0  9.666667

#### Safe Zone Places ranking during Night

In [54]:
open_night =  route_data['open_at_night']=='Yes'

In [57]:
route_data_night = route_data[open_night]

In [58]:
route_data_night

Public Place  Distance(km)  Duration(mins)  \
0  Women Police Station           0.8               5   
1        Police Station           1.0               7   
6         Moving Patrol           0.9               9   
7         24/7 Hospital           1.4               8   
9      24/7 Petrol Bunk           1.4              10   

   User Rating(1-Poor,5-Excellent) open_at_night  Weightage  Distance_Rank  \
0                                5           Yes        1.0            1.0   
1                                5           Yes        0.9            3.0   
6                                5           Yes        0.9            2.0   
7                                5           Yes        1.0           10.5   
9                                5           Yes        1.0           10.5   

   Duration_Rank  Rating_Rank     Score  
0            1.0          3.0  1.666667  
1            3.0          3.0  3.000000  
6            6.0          3.0  3.666667  
7            4.5          3.0  6.000000  
9            7.5          3.0  7.000000